Move to our own scheduling

grote opened the pull request:
Closes #583, #629, #590

grote committed:
Move APK backup from BackupCoordinator to new ApkBackupManager

This is a preparation for doing APK backup ourselves in a worker and not hacked into the backup transport. The latter was prone to timeouts by the AOSP backup API. With a worker, we have a bit more control and can schedule backups ourselves.

grote committed:
Make 'Backup now' action use AppBackupWorker

grote committed:
Improve backup notification

grote committed:
Expose time of next backup in UI

to help debugging scheduling issues

grote committed:
Migrate to own backup scheduling

grote committed:
Guard against BadParcelableException when getting app list

hopefully something rare, but it just happened to me while testing.

It seems it happens when there are many apps installed (>500) and the app list is open while a backup happens. Then, we keep reloading the list and hammer the package manager hard which it seems can't handle it. It does recover on its own though.

grote committed:
Respect when worker was stopped

and log worker ID as well as object, because we've seen two scheduled workers running at the same time, requesting a backup at the same time. This should not happen.

grote committed:
During restore, show apps without APK as failed

Previously, we backed up APKs of apps we could not back up (even if APK backup was disabled) so the user had a chance to get at least the apps back when restoring. Now, it is enough to record metadata about the app and the user will be able to manually install the app. The install apps step won't be skipped anymore.

grote committed:
Properly schedule/cancel backup workers when backup destination changes

When the user changes to USB storage, we need to cancel current schedulings, because the storage is not always available (maybe can use a trigger URI?). And if moving to a non-USB storage, we need to schedule backups again.

Unfortunately, there are two places in the code where we handle storage location changes. Ideally, those get unified at some point.

grote committed:
Use WorkInfo for determining if a backup is already running

Backup and restore is not possible when a backup is running. We used to check notifications for this, but now can use WorkManager's WorkInfo which should be more reliable.

Also, we used to prevent the "Backup now" action when app backup was disabled. But the user may want to do a storage backup. This is now possible.

grote committed:
Only reschedule next app backup when not on USB storage

Currently, after a manual run, we need to schedule the background backups again, because the scheduling gets lost. However, we need to be careful not to do that when the backup destination is on removable storage. Then we don't want to run.

grote committed:
try more than once to upload metadata after APK backups

Failure to upload metadata after backup up APKs can be critical and flaky I/O can make it fail, so we try again.

grote committed:
Expose scheduling options in the UI

grote committed:
Allow backups in metered network, if user wants that

grote committed:
Stop running instrumentation tests for SDK 33

grote committed:
Don't retry instrumentation tests

grote committed:
Request @pm@ backup after initialization

to avoid a 2nd restore set being used.

This also changes the initialization behavior to only create the restore set folder and upload the metadata only when we actually need to. This way, double inits are not creating new restore sets on the backup destination.

grote committed:
Initialize backup, when enabling it

For ApkBackup, we need to be initialized. If the system starts with app backup off, we would not initialize which would lead to issues when backing up the APKs.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/UsbIntentReceiver.kt:
```diff
@@ -63,9 +63,7 @@ class UsbIntentReceiver : UsbMonitor() {
-            Thread {
-                requestBackup(context)
-            }.start()
+            AppBackupWorker.scheduleNow(context, reschedule = false)
```
Maybe this should use the feature flag?

grote replied:
Why a feature flag? This is the same behavior as before. When the user plugs in the flash drive and sufficient time has passed, it requests a backup *now*.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/App.kt:
```diff
@@ -115,6 +126,25 @@ open class App : Application() {
+            AppBackupWorker.schedule(applicationContext, settingsManager, UPDATE)
+        }
+        // cancel old D2D worker
+        WorkManager.getInstance(this).cancelUniqueWork("APP_BACKUP")
```
Should this be a constant in `AppBackupWorker`?

grote replied:
This is migratory code, so I had decided against it. This is the only time we'll ever going to use it. This is for people that had tried the d2d feature flag and used a worker before. This is now superseded by the new workers.

The real unique work IDs we are still using have constants.

theimpulson replied:
Marking as resolved assuming we are fine with this here.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/metadata/MetadataManager.kt:
```diff
@@ -192,9 +174,60 @@ internal class MetadataManager(
+     */
+    @Synchronized
+    @Throws(IOException::class)
+    internal fun onPackageDoesNotGetBackedUp(
```
Maybe just `updatePackageState`? This indicates what it does in the function name, rather than where it's being called from.

grote replied:
I am a bit torn here, as all the other methods are also using names related to the purpose of the call rather the implementation (e.g. `onPackageBackupError()`).

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/metadata/MetadataManager.kt:
```diff
@@ -192,9 +174,60 @@ internal class MetadataManager(
+        packageInfo: PackageInfo,
+        packageState: PackageState,
+    ) = modifyCachedMetadata {
+        val packageName = packageInfo.packageName
```
```
        metadata.packageMetadataMap.getOrPut(packageInfo.packageName) {
            PackageMetadata(
                time = 0L,
                state = packageState,
                system = packageInfo.isSystemApp(),
            )
        }.state = packageState
```

grote replied:
good call. I used getOrPut() here and in the other places where it made sense.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/SettingsActivity.kt:
```diff
@@ -36,6 +37,19 @@ class SettingsActivity : RequireProvisioningActivity(), OnPreferenceStartFragmen
+                val title = getString(R.string.storage_check_fragment_backup_title)
+                showFragment(StorageCheckFragment.newInstance(title), true, tag)
+            } else {
+                val f = supportFragmentManager.findFragmentByTag(tag)
```
Maybe `fragment` as variable name, as single letter variable names can be considered confusing?

grote replied:
Sure!

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/SettingsActivity.kt:
```diff
@@ -36,6 +37,19 @@ class SettingsActivity : RequireProvisioningActivity(), OnPreferenceStartFragmen
+                showFragment(StorageCheckFragment.newInstance(title), true, tag)
+            } else {
+                val f = supportFragmentManager.findFragmentByTag(tag)
+                if (f != null && f.isVisible) supportFragmentManager.popBackStack()
```
Could be `if(fragment?.isVisible == true)`.

grote replied:
Not sure which is more readable, but ok.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt:
```diff
@@ -244,10 +255,41 @@ class SettingsFragment : PreferenceFragmentCompat() {
+        }
+    }
+
+    private fun setAppBackupSchedulingSummary(workInfo: WorkInfo?) {
```
This function is on the larger side, maybe could use some documentation?

grote replied:
Alright, added.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/SettingsManager.kt:
```diff
@@ -134,13 +146,24 @@ class SettingsManager(private val context: Context) {
         return prefs.getBoolean(PREF_KEY_BACKUP_APK, true)
     }
 
+    val backupFrequencyInMillis: Long
```
I assume we plan on exposing this to the user?

grote replied:
It is already exposed and configurable via `SchedulingFragment`.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/ExpertSettingsFragment.kt:
```diff
@@ -16,10 +16,10 @@ class ExpertSettingsFragment : PreferenceFragmentCompat() {
-        viewModel.onLogcatUriReceived(uri)
-    }
+    private val createFileLauncher =
+        registerForActivityResult(CreateDocument("text/plain")) { uri ->
```
`org.apache.http.entity.TEXT_PLAIN.mimeType` maybe or something similar

grote replied:
I found `com.google.android.mms.ContentType.TEXT_PLAIN` and now use that, but not sure if that's better than just leaving the string in.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt:
```diff
@@ -244,10 +255,41 @@ class SettingsFragment : PreferenceFragmentCompat() {
+            backupScheduling.summary = getString(R.string.settings_backup_status_next_backup_usb)
+            return
+        }
+        if (workInfo == null) return
```
Do we need to clear any existing state?

grote replied:
Good question. I don't *think* we need to, because it should only ever be `null` when the workInfo wasn't loaded. Afterwards, we *should* always get something non-null here.

stevesoltys commented on app/src/main/java/com/stevesoltys/seedvault/transport/backup/BackupCoordinator.kt:
```diff
@@ -120,18 +116,15 @@ internal class BackupCoordinator(
-        plugin.getMetadataOutputStream(token).use {
-            metadataManager.onDeviceInitialization(token, it)
-        }
+        startNewRestoreSet()
```
We are starting two restore sets, intentionally I suppose? Maybe a comment on why?

grote replied:
No, I think this is an oversight on my part. I'll take a look at that separately, as this is more delicate.

grote replied:
Yeah, so this wasn't correct. Well spotted. Now, we are only doing this once. I tested it and it should work.

stevesoltys reviewed (dismissed):
Mainly just nitpicking.

grote committed:
Address review feedback

grote committed:
Try to close system dialogs of emulator

grote committed:
Report total number of user apps when showing final notification

Before, we showed the number of apps we requested the backup for which in case of non-d2d may be much lower than the number of installed apps.

In the future we may decide to also include certain system apps in that count.

grote committed:
Do live-counting of backed up apps for success notification

Previously, we asked the MetadataManager which also includes historic data and may provide misleading totals.

theimpulson reviewed (approved):
Looks good to me. Feel free to merge when ready.
